### ISSUE TITLE
Proposal: generic write() method

### ISSUE THREAD

**Issue by jmcnamara** [jmcnamara/rust_xlsxwriter#16]
It would be possible to have a generic `write()` method using a trait that could call the type specific `write_*()` method such as  `write_number()`. This is a feature in the [Python version](https://xlsxwriter.readthedocs.io/worksheet.html#worksheet-write) of the library.

I'll leave this here as a reminder to myself and to see if it gets an +1s for usefulness or desirability.

**Comment by moutikabdessabour** [comment]
Can I take a shot at this?

**Comment by jmcnamara** [comment]
@moutikabdessabour At this stage I'm mainly asking if people think this would be useful. Do you think it would be useful?

Personally I'm currently more against this than for the reasons below but I'm willing to take input. 

**Pros**:

- Makes writing examples easier. 

**Cons**:

- The similar `write()` method in the Python and Perl versions tend to generate bug reports when users try to write arbitrary sting or number like data. People also assume that writing something like "2003-01-01" should become a date in Excel. 
- There would be a slight performance overhead for strings since they would need to be checked to see if they are formulas or urls

**Comment by moutikabdessabour** [comment]
For the second con I think implementing a Formula/Url type would remove the overhead. I wrote an app using this library and while it is great the only issue I had is the verbosity of the function calls. I am a 100% for this feature

**Comment by jmcnamara** [comment]
> I wrote an app using this library and while it is great the only issue I had is the verbosity of the function calls.

Yes, that is my feeling too. There are currently 24 `write_*()` methods. That is just too many.

It could be mapped down to 5 like this:

| Current                                     | New                                 |
| ------------------------------------------- | ----------------------------------- |
| `write_number()`                            | `write()`                           |
| `write_string()`                            | `write()`                           |
| `write_url()`                               | `write()`                           |
| `write_boolean()`                           | `write()`                           |
| `write_formula()`                           | `write()`                           |
| `write_rich_string()`                       | `write()`                           |
| `write_blank()`                             | `write()`                           |
| `write_time()`                              | `write()`                           |
| `write_date()`                              | `write()`                           |
| `write_datetime()`                          | `write()`                           |
| `write_dynamic_formula()`                   | `write()`                           |
|                                             |                                     |
| `write_number_with_format()`                | `write_with_format()`               |
| `write_string_with_format()`                | `write_with_format()`               |
| `write_url_with_format()`                   | `write_with_format()`               |
| `write_boolean_with_format()`               | `write_with_format()`               |
| `write_formula_with_format()`               | `write_with_format()`               |
| `write_rich_string_with_format()`           | `write_with_format()`               |
| `write_dynamic_formula_with_format()`       | `write_with_format()`               |
|                                             |                                     |
| `write_array_formula()`                     | `write_array_formula()`             |
| `write_dynamic_array_formula()`             | `write_array_formula()`             |
|                                             |                                     |
| `write_array_formula_with_format()`         | `write_array_formula_with_format()` |
| `write_dynamic_array_formula_with_format()` | `write_array_formula_with_format()` |
|                                             |                                     |
| `write_url_with_options()`                  | Same                                |
| `write_url_with_text()`                     | Same, or `write_url_with_options()` |



With some caveats/additions:

- It will need the addition of a new `Url` type.
- It will need the addition of a new `Formula` type that can be set to array or dynamic array (if needed)
- Default DateTime, Date and Time formats for datetime()/date()/time() variants without a user specified format since a date without a format in Excel is just a number.

Also a `IntoExcelType` (or similar) trait coupled with a generic `write()` would have the advantage of allowing mapping of other arbitrary user types in a clean way.

Now, that I've mapped it out like this it seems like the right thing to do (even if it is another major API break). I might add it after I finish the current work on charts.

**Comment by moutikabdessabour** [comment]
Gonna take a jab at this later today. WIll probably have something ready tomorrow night.

**Comment by jmcnamara** [comment]
Not to deter you from doing a prototype. I just to flag that it will needs a lot of doc and example refactoring and I'd prefer to do that myself.

**Comment by GlennMm** [comment]
i would love the feature as it would enable me for example to write one function to write json data in array format (structured differently) with having to write many functions for a specific struct to write json to excel. If you dim this feature as unnecessary then can you link me a rs lib that enable me todo so if its available.

**Comment by jmcnamara** [comment]
@GlennMm At this point I'm definitely going to add it. Either in the next release or a separate release just after.

**Comment by GlennMm** [comment]
> @GlennMm At this point I'm definitely going to add it. Either in the next release or a separate release just after.

Thank you @jmcnamara that would awesome, looking forward to this feature

**Comment by jmcnamara** [comment]
Also, if the Format was treated as an `Option` it would be possible to have just one method that needs to be implemented for the trait:

```rust
impl IntoExcelData for UnixTime {
    fn write<'a>(
        self,
        worksheet: &'a mut Worksheet,
        row: RowNum,
        col: ColNum,
        format: Option<&'a Format>,
    ) -> Result<&'a mut Worksheet, XlsxError> {
        // Convert the Unix time to an Excel datetime.
        let datetime = 25569.0 + (self.seconds / (24 * 60 * 60)) as f64;

        match format {
            Some(format) => {
                // Write the date with the user supplied format.
                worksheet.write_number_with_format(row, col, datetime, format)
            }
            None => {
                // Create a default date format.
                let format = Format::new().set_num_format("yyyy-mm-dd");
                worksheet.write_number_with_format(row, col, datetime, &format)
            }
        }
    }
}
```

This may, or may not, be more confusing to the end user. Internally it works out a lot cleaner.

**Comment by jmcnamara** [comment]
@Fight-For-Food Any thoughts on the interface or implementation of this feature? You had some good input on other generic/non-generic code.

**Comment by Fight-For-Food** [comment]
> @Fight-For-Food Any thoughts on the interface or implementation of this feature? You had some good input on other generic/non-generic code.

I think that
`fn write(self, worksheet: &mut Worksheet, row: RowNum, col: ColNum) -> Result<&mut Worksheet, XlsxError>{....}`
is ok. It is convenient to use. And there is no need to make in more complex with extra-argument `format: Option<&'a Format>`
For some more comprehensive thoughts i need more time. Maybe i will have some at this week. But can't make a promise about it.

**Comment by jmcnamara** [comment]
> And there is no need to make in more complex with extra-argument `format: Option<&'a Format>`

It may not be clear from the snippet above but the addition of the `Option<>` means that there is only one trait funtion that needs to be implemented by the end user. However, it would probably be confusing and the version on main with `IntoExcelData::write()` and `IntoExcelData::write_with_format()` mimics the worksheet APIs may be simpler.

> Maybe i will have some at this week. But can't make a promise about it.

No problem if you can't get to it.

**Comment by GlennMm** [comment]
thank you for implementing this, it really help make my work much easier. looking to forward to be start contributing to this repo. so have started using this feature i haven't seen any problems with it so far

**Comment by jmcnamara** [comment]
Just a heads up that the initial version of this is upstream in crates.io in version 0.27.0: https://crates.io/crates/rust_xlsxwriter

I'll iterate and improve on this in the next few releases.

**Comment by jmcnamara** [comment]
Having taken a small detour into charts I've come back to complete this.

Some tasks, not in order:

- Use `impl Into<String>` instead of &str in method parameters where possible/sensible.
- Implement generic `write_row()` and `write_column()` methods.
- Add types for `Url` and `Formula` and add them to generic `write()`.

**Comment by jmcnamara** [comment]
> Implement generic write_row() and write_column() methods.

I've added a first pass at this to main. You can now write arrays of data types that implement `IntoExcelData` like this:

```rust
use rust_xlsxwriter::{Workbook, XlsxError};

fn main() -> Result<(), XlsxError> {
    // Create a new Excel file object.
    let mut workbook = Workbook::new();

    // Add a worksheet to the workbook.
    let worksheet = workbook.add_worksheet();

    // Some array data to write.
    let numbers = [1, 2, 3, 4, 5];
    let words = ["Hello"; 5];

    worksheet.write_column(0, 0, &numbers)?;
    worksheet.write_column(0, 1, &words)?;

    worksheet.write_row(0, 3, &numbers)?;
    worksheet.write_row(1, 3, &words)?;

    // Save the file to disk.
    workbook.save("arrays.xlsx")?;

    Ok(())
}
```

Output:

![screenshot](https://user-images.githubusercontent.com/94267/236587184-0be49eca-eb33-4ede-b71d-0a89ff5f499c.png)

It is just some syntactic sugar around a loop but it is somewhat popular in the Python version.

It currently has an additional deference/copy and I'm not sure if that can be avoided. Feedback welcome on that or anything else.

**Comment by adriandelgado** [comment]
I gave [some feedback](https://github.com/jmcnamara/rust_xlsxwriter/commit/59c9be75b382345c567a4a1769322c4ce7c3feea#r112069140) on the last commit. I think that feedback applies on more parts of the codebase.

**Comment by jmcnamara** [comment]
@adriandelgado 

> I think the Copy trait bound could be avoided by using the [IntoIterator](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html) trait.

I initially tried, and failed, to get that working. :-( With my morning brain and 2 cups of coffee I managed to understand/fix it. You can have a look.

> I gave [some feedback](https://github.com/jmcnamara/rust_xlsxwriter/commit/59c9be75b382345c567a4a1769322c4ce7c3feea#r112069140) on the last commit. I think that feedback applies on more parts of the codebase.

I think I understood those comments but maybe not so let's check. What I think you are saying is that the API/library should avoid taking any references to data that will then be cloned and should instead use interfaces like `Into<String>` or `IntoInterator`. Is that correct or am I missing something else?

**Comment by jmcnamara** [comment]
I've also added `write_row_matrix()` and `write_column_matrix()` methods for 2D arrays/iterators of data.

**Comment by adriandelgado** [comment]
> I think I understood those comments but maybe not so let's check. What I think you are saying is that the API/library should avoid taking any references to data that will then be cloned and should instead use interfaces like `Into<String>` or `IntoInterator`. Is that correct or am I missing something else?

That's right. The caller of the function has the responsability to clone if needed. Sometimes the caller already has owned data and knows it wont need it anymore, no clone needed in that case.

**Comment by jmcnamara** [comment]
I've implemented a `Formula` type so that it can be used in the generic `write()`:

```rust
    worksheet.write(0, 0, Formula::new("=1+2+3"))?;
    worksheet.write(1, 0, Formula::new("=SIN(PI())"))?;
```

However, I'm not sure how much of a help this is yet.

**Comment by jmcnamara** [comment]
I had hoped to be able to use [Trait Objects](https://doc.rust-lang.org/book/ch17-02-trait-objects.html) and the newer array writing methods to simulate something like writing a heterogeneous vec of vectors like a dataframe.  

Something like this:

```rust
    let data: Vec<Vec<Box<dyn IntoExcelData>>> = vec![
        vec![
            Box::new("East"),
            Box::new("West"),
            Box::new("North"),
            Box::new("South"),
        ],
        vec![
            Box::new(40),
            Box::new(40),
            Box::new(20),
            Box::new(30)],
        vec![
            Box::new(30),
            Box::new(30),
            Box::new(50),
            Box::new(30)],
        vec![
            Box::new(Formula::new("=B1+C1")),
            Box::new(Formula::new("=B2+C2")),
            Box::new(Formula::new("=B3+C3")),
            Box::new(Formula::new("=B4+C4")),
        ],
    ];
```

However, I've failed in every attempt so far. Examples like this with a simpler data structure:

```rust
use rust_xlsxwriter::{Formula, IntoExcelData, Workbook, XlsxError};

fn main() -> Result<(), XlsxError> {
    // Create a new Excel file object.
    let mut workbook = Workbook::new();

    // Add a worksheet to the workbook.
    let worksheet = workbook.add_worksheet();

    let data: Vec<Box<dyn IntoExcelData>> = vec![
        Box::new("East"),
        Box::new(50),
        Box::new(30),
        Box::new(Formula::new("=B1+C1")),
    ];

    /* The following doesn't work.

        --> examples/app_trait_array.rs:39:9
        |
     40 |         item.write(worksheet, row, 1);
        |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the size of `dyn IntoExcelData`
                                                cannot be statically determined

    */
    let mut row = 1;
    for item in data {
        item.write(worksheet, row, 1);
        row += 1;
    }

    // Save the file to disk.
    workbook.save("arrays.xlsx")?;

    Ok(())
}
```

Or this:

```rust
    let mut row = 1;
    for item in data {
        worksheet.write(row, 0, item)?;
        row += 1;
    }
```

Which gives errors like this:

```rust
   --> examples/app_trait_array3.rs:28:33
    |
28  |         worksheet.write(row, 0, item)?;
    |                   -----         ^^^^ the trait `IntoExcelData` is not implemented 
                                          for `Box<dyn IntoExcelData>`
    |                   |
    |                   required by a bound introduced by this call
    |
    = help: the following other types implement trait `IntoExcelData`:
```

Anyone see a way that trait objects for the `IntoExcelData` trait might be handled?

**Comment by jmcnamara** [comment]
See also this example of writing a Polars dataframe to Excel using `rust_xlsxwriter`: #39

**Comment by jmcnamara** [comment]
These generic methods and types are now available in release v0.39.0.
